### PR TITLE
Add pre-commit check for docs cross-references

### LINKED ISSUE #5317: Add pre-commit check for sphinx cross-references
Now that we have versioned docs, we can't use urls for internal links if we want each version to point to themselves. I think we should add a pre-commit check to ensure there are no links pointing to `docs.pymc.io` anywhere inside the `docs/source/` folder.

I don't know pre-commit but it seems like this should be easy for someone who does know their way around it. cc @MarcoGorelli if you agree with this, can you label it beginner friendly?

### PR DESCRIPTION
**Pull request by symeneses**
It closes #5317 in #DataUmbrellaPyMCSprint

Adds a hook to check for  `docs.pymc.io` inside `docs/source/`.

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] [linting/style checks have been run](https://docs.pymc.io/en/latest/contributing/python_style.html)

This PR should be merged after the existing links are reviewed. The hook is detecting these references to be fixed:

- [x] docs/source/contributing/developer_guide.rst:65:https://docs.pymc.io/Probability\_Distributions.html#using-pymc-distributions-without-a-model).
- [x] docs/source/contributing/developer_guide.rst:113:https://docs.pymc.io/Probability\_Distributions.html#custom-distributions
- [x] docs/source/contributing/developer_guide.rst:377:   transformation <https://docs.pymc.io/notebooks/api_quickstart.html?highlight=chain%20transformation>`__).
- [x] docs/source/contributing/developer_guide.rst:693:method <https://docs.pymc.io/notebooks/sampling_compound_step.html>`__.
- [x] docs/source/contributing/developer_guide.rst:696:RVs <https://docs.pymc.io/notebooks/sampling_compound_step.html?highlight=compoundstep#Specify-compound-steps>`__.
- [x] docs/source/contributing/developer_guide.rst:714:doc <https://docs.pymc.io/notebooks/sampling_compound_step.html#Specify-compound-steps>`__.
- [x] docs/source/contributing/developer_guide.rst:913:    -  `Using “black box” likelihood function by creating a custom Aesara Op <https://docs.pymc.io/notebooks/blackbox_external_likelihood.html>`__
- [x] docs/source/contributing/developer_guide_implementing_distribution.md:26:Before creating a new `RandomVariable` make sure that it is not offered in the [Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions).
- [x] docs/source/contributing/release_checklist.md:25:+ Make sure the new version appears at the website and that [`docs.pymc.io/en/stable`](https://docs.pymc.io/en/stable) points to it.
- [x] docs/source/glossary.md:14:  Choosing which function or method implementation to use based on the type of the input variables (usually just the first variable). For some examples, see Python's documentation for the [singledispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) decorator.
- [x] docs/source/learn/examples.md:13:- https://docs.pymc.io/en/stable/Probability_Distributions.html
- [x] docs/source/learn/examples.md:16:- https://docs.pymc.io/en/stable/PyMC3_and_Theano.html

### CODE REVIEW

**Comment by OriolAbril** [comment]
Thanks! Will review asap

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -60,3 +60,10 @@ repos:
       entry: from \.[\.\w]* import
       types: [python]
       language: pygrep
+    - id: no-internal-links
+      name: Check no internal links are in the docs
+      description: 'A quick check for the `docs.pymc.io` inside `docs/source/`'
+      files: ^docs/source/
+      entry: 'docs.pymc.io'
+      language: pygrep
+      types_or: [markdown, rst]
```
Comment (OriolAbril) [inline .pre-commit-config.yaml]: This should also include ipynb files, there are some jupyter notebooks that are part of the docs and should follow the same rules

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -60,3 +60,10 @@ repos:
       entry: from \.[\.\w]* import
       types: [python]
       language: pygrep
+    - id: no-internal-links
+      name: Check no internal links are in the docs
+      description: 'A quick check for the `docs.pymc.io` inside `docs/source/`'
+      files: ^docs/source/
+      entry: 'docs.pymc.io'
```
Comment (OriolAbril) [inline .pre-commit-config.yaml]: is it possible to also check for aesara and arviz docs?

**Comment by OriolAbril** [comment]
commenting with what I believe are the right sphinx cross references for these links, but we'll still need to check the preview to make sure they work.

```
docs/source/contributing/developer_guide.rst:10:`PyMC <https://docs.pymc.io/>`__ is a Python package for Bayesian

:doc:`PyMC <index>`
```

```
docs/source/contributing/developer_guide.rst:17:quickstart <https://docs.pymc.io/notebooks/api_quickstart.html>`__. A more accessible, user facing deep introduction can be found in

:ref:`quickstart <pymc_overview>`
```

```
docs/source/contributing/developer_guide.rst:24:the `documentation <https://docs.pymc.io/Probability_Distributions.html>`__. The source
docs/source/contributing/developer_guide.rst:66:https://docs.pymc.io/Probability\_Distributions.html#using-pymc-distributions-without-a-model).
docs/source/contributing/developer_guide.rst:115:https://docs.pymc.io/Probability\_Distributions.html#custom-distributions
```

No idea about this one, will have to check where it was linking originally, this page is now gone from the docs I think :thinking: 

```
docs/source/contributing/developer_guide.rst:70:<https://docs.pymc.io/api/distributions.html>`__), which divided into:

:ref:`api_distributions`
```

This also needs to create the `api_distributions` anchor in this file: https://github.com/pymc-devs/pymc/blob/main/docs/source/api/distributions.rst. That would be adding `.. _api_distributions:` in the first line (before the title)

```
docs/source/contributing/developer_guide.rst:323:   <https://docs.pymc.io/notebooks/api_quickstart.html#Automatic-transforms-of-bounded-RVs>`__.
docs/source/contributing/developer_guide.rst:329:   `doc <https://docs.pymc.io/notebooks/api_quickstart.html#Transformed-distributions-and-changes-of-variables>`__):
docs/source/contributing/developer_guide.rst:380:   transformation <https://docs.pymc.io/notebooks/api_quickstart.html?highlight=chain%20transformation>`__).
```

These need changes to the api quickstart notebook. I can do these changes and then I'll update with the references that should be used.

```
docs/source/contributing/developer_guide.rst:696:method <https://docs.pymc.io/notebooks/sampling_compound_step.html>`__.
docs/source/contributing/developer_guide.rst:699:RVs <https://docs.pymc.io/notebooks/sampling_compound_step.html?highlight=compoundstep#Specify-compound-steps>`__.
docs/source/contributing/developer_guide.rst:717:doc <https://docs.pymc.io/notebooks/sampling_compound_step.html#Specify-compound-steps>`__.

:ref:`nb:sampling_compound_step`  # or variations
```

this also needs updates to the notebook, will get back about them in a bit.

```
docs/source/contributing/developer_guide.rst:916:    -  `Using “black box” likelihood function by creating a custom Aesara Op <https://docs.pymc.io/notebooks/blackbox_external_likelihood.html>`__

:ref:`nb:blackbox_external_likelihood`
```

```
docs/source/contributing/release_checklist.md:25:+ Make sure the new version appears at the website and that [`docs.pymc.io/en/stable`](https://docs.pymc.io/en/stable) points to it.
```

This is actually a valid occurrence of the website :scream: can we skip this file from the checks maybe?

```
docs/source/index.md:158:Examples <https://docs.pymc.io/projects/examples/en/latest/>

:ref:`nb:index`
```

```
docs/source/learn/examples.md:6:- https://docs.pymc.io/en/stable/pymc-examples/examples/generalized_linear_models/GLM-linear.html

:ref:`nb:GLM-linear`
```

```
docs/source/learn/examples.md:9:- https://docs.pymc.io/en/stable/pymc-examples/examples/case_studies/multilevel_modeling.html

{ref}`nb:multilevel_modeling`  # note this is in a markdown file, so curly brackets instead of colons
```

```
docs/source/learn/examples.md:12:- https://docs.pymc.io/en/stable/pymc-examples/examples/diagnostics_and_criticism/model_comparison.html

{ref}`model_comparison`  # this has been moved from the examples so there should be no nb:
```

```
docs/source/learn/examples.md:16:- https://docs.pymc.io/en/stable/Probability_Distributions.html
docs/source/learn/examples.md:19:- https://docs.pymc.io/en/stable/PyMC3_and_Theano.html
```

I think these two files are gone :thinking: 

```
docs/source/learn/examples.md:23:- https://docs.pymc.io/en/stable/pymc-examples/examples/diagnostics_and_criticism/posterior_predictive.html

{ref}`posterior_predictive`   # also moved from examples, no nb:
```

```
docs/source/learn/examples.md:26:- https://docs.pymc.io/en/stable/Gaussian_Processes.html

{mod}`pymc.gp`
```

File: .pre-commit-config.yaml (hunk 2 above)
Comment (symeneses) [reply .pre-commit-config.yaml]: Almost everything is possible 😉 . Should I include `aesara.readthedocs.io `and `arviz-devs.github.io`?

Comment (OriolAbril) [reply .pre-commit-config.yaml]: Yes, that would be amazing, if it's not too difficult, can you add all websites in the intersphinx mapping?

https://github.com/pymc-devs/pymc/blob/main/docs/source/conf.py#L179

File: .pre-commit-config.yaml (hunk 1 above)
Comment (symeneses) [reply .pre-commit-config.yaml]: Done ✔️

File: .pre-commit-config.yaml (hunk 2 above)
Comment (symeneses) [reply .pre-commit-config.yaml]: Done ✔️

File: docs/source/404.md [hunk 3]
```diff
@@ -28,4 +28,4 @@ do so we have moved some files around and we have
 modified the base url in order to support multi version docs
 (you'll see the version switcher at the bottom right of the page).
 
-Back to {{ '[documentation homepage](https://docs.pymc.io/en/{}/)'.format(version_slug) }}
```
Comment (OriolAbril) [inline docs/source/404.md]: the 404 file is special and should html link, we should exclude this file from the pre-commit hook.

**Comment by OriolAbril** [comment]
files to be excluded:

```
404
contributing/release_checklist
contributing/developer_guide
learn/examples.md
```

**Comment by OriolAbril** [comment]
docs/source/contributing/developer_guide_implementing_distribution.md:26:Before creating a new RandomVariable make sure that it is not offered in the [Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions).

`` {mod}`numpy.random` ``

**Comment by OriolAbril** [comment]
docs/source/glossary.md:14: Choosing which function or method implementation to use based on the type of the input variables (usually just the first variable). For some examples, see Python's documentation for the [singledispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) decorator.

`` {func}`~functools.singledispatch` ``

File: docs/source/404.md (hunk 3 above)
Comment (symeneses) [reply docs/source/404.md]: It's now how it was before

**Comment by symeneses** [comment]
> files to be excluded:
> 
> ```
> 404
> contributing/release_checklist
> contributing/developer_guide
> learn/examples.md
> ```

After ignoring the files the check is passing.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -60,3 +60,32 @@ repos:
...
       language: pygrep
+    - id: no-internal-links
+      name: Check no internal links are in the docs
+      description: >-
+        'A quick check for the links in the intersphinx '
+        'mapping inside `docs/source/`'
+        'See docs/source/conf.py for more information.'
+      files: ^docs/source/
+      # Files that should be ignored:
+      # - 404.md
```
Comment (OriolAbril) [inline .pre-commit-config.yaml:72]:
```suggestion
      # - 404.md
      # - contributing_release_checklist.md
```

One of the steps in the release checklist is checking that the documentation has been built correctly and is served and available on readthedocs. It has the url to be checked written explicitly to make sure the task is clear even if docs or cross-referencing doesn't work as expected.

File: docs/source/contributing/developer_guide.rst [hunk 5]
```diff
@@ -7,25 +7,24 @@
...
-quickstart <https://docs.pymc.io/notebooks/api_quickstart.html>`__. A more accessible, user facing deep introduction can be found in
+introduction can be found in the :ref:`API quickstart <pymc_overview>`. A more accessible, user facing deep introduction can be found in
 `Peadar Coyle's probabilistic programming primer <https://github.com/springcoil/probabilisticprogrammingprimer>`__
 
 Distribution
 ------------
 
 A high-level introduction of ``Distribution`` in PyMC can be found in
-the `documentation <https://docs.pymc.io/Probability_Distributions.html>`__. The source
+the :doc:`documentation <api:distributions>`. The source
```
Comment (OriolAbril) [inline docs/source/contributing/developer_guide.rst]:
```suggestion
the :ref:`documentation <api_distributions>`. The source
```

File: docs/source/contributing/developer_guide.rst [hunk 6]
```diff
@@ -7,25 +7,24 @@
...
 code of the probability distributions is nested under
-`pymc/distributions <https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/>`__,
-with the ``Distribution`` class defined in `distribution.py
-<https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/distribution.py#L23-L44>`__.
+:doc:`pymc/distributions <api:distributions>`,
```
Comment (OriolAbril) [inline docs/source/contributing/developer_guide.rst]:
```suggestion
:ref:`pymc/distributions <api_distributions>`,
```

File: docs/source/contributing/developer_guide.rst [hunk 7]
```diff
@@ -7,25 +7,24 @@
...
+with the ``Distribution`` class defined in :class:`distribution.py
+pymc.Distribution`.
```
Comment (OriolAbril) [inline docs/source/contributing/developer_guide.rst]:
```suggestion
with the :class:`~pymc.Distribution` class defined in ``distribution.py``.
```

File: docs/source/index.md [hunk 8]
```diff
@@ -155,7 +155,7 @@ See [Google Scholar](https://scholar.google.de/scholar?oi=bibs&hl=en&authuser=1&
 
 installation
 learning
-Examples <https://docs.pymc.io/projects/examples/en/latest/>
+Examples <nb:index>
```
Comment (OriolAbril) [inline docs/source/index.md]: this one is not working like this :/. The preview is missing examples from the navbar. Can you undo the change and add it to the list of excluded files?

File: docs/source/learn/examples.md [hunk 9]
```diff
@@ -3,14 +3,11 @@
 
 TODO add "core" notebooks:
 
-- https://docs.pymc.io/en/stable/pymc-examples/examples/generalized_linear_models/GLM-linear.html
-  - https://github.com/pymc-devs/pymc-examples/blob/main/examples/generalized_linear_models/GLM-linear.ipynb
+- {ref}`nb:GLM-linear`
```
Comment (OriolAbril) [inline docs/source/learn/examples.md]:
```suggestion
- {ref}`GLM-linear`
```

there are a handful of notebooks in the versioned docs that need to be referenced without the `nb:` as they are not part of the example gallery website

File: docs/source/learn/examples.md [hunk 10]
```diff
@@ -3,14 +3,11 @@
...
 
-- https://docs.pymc.io/en/stable/pymc-examples/examples/case_studies/multilevel_modeling.html
-  - https://github.com/pymc-devs/pymc-examples/blob/main/examples/case_studies/multilevel_modeling.ipynb
+- {ref}`nb:multilevel_modeling`
 
-- https://docs.pymc.io/en/stable/pymc-examples/examples/diagnostics_and_criticism/model_comparison.html
-  - https://github.com/pymc-devs/pymc-examples/blob/main/examples/diagnostics_and_criticism/model_comparison.ipynb
+- {ref}`nb:model_comparison`
```
Comment (OriolAbril) [inline docs/source/learn/examples.md]:
```suggestion
- {ref}`model_comparison`
```

File: docs/source/learn/examples.md [hunk 11]
```diff
@@ -20,11 +17,9 @@ TODO add "core" notebooks:
   update needed: from "pymc and theano" to "pymc and aesara" tutorial (not just auto-renaming from theano to aesara but showing correct use). Very related to migration guide as well, though the migration guide already covers it.
   - https://github.com/pymc-devs/pymc/blob/37ba9a3e3a19b738f48cb30007f4d70c33bdd0f6/docs/source/PyMC_and_Aesara.rst
 
-- https://docs.pymc.io/en/stable/pymc-examples/examples/diagnostics_and_criticism/posterior_predictive.html
-  - https://github.com/pymc-devs/pymc-examples/blob/2002ebd815a199be89b011039906b197bca42361/examples/diagnostics_and_criticism/posterior_predictive.ipynb
+- {ref}`nb:posterior_predictive`
```
Comment (OriolAbril) [inline docs/source/learn/examples.md]:
```suggestion
- {ref}`posterior_predictive`
```

File: docs/source/learn/usage_overview.rst [hunk 12]
```diff
@@ -6,7 +6,7 @@ Usage Overview
 
 For a detailed overview of building models in PyMC, please read the appropriate sections in the rest of the documentation. For a flavor of what PyMC models look like, here is a quick example.
 
-First, let's import PyMC and `ArviZ <https://arviz-devs.github.io/arviz/>`__ (which handles plotting and diagnostics):
+First, let's import PyMC and :doc:`ArviZ <arviz>` (which handles plotting and diagnostics):
```
Comment (OriolAbril) [inline docs/source/learn/usage_overview.rst]:
```suggestion
First, let's import PyMC and :doc:`ArviZ <arviz:index>` (which handles plotting and diagnostics):
```

File: docs/source/index.md (hunk 8 above)
Comment (symeneses) [reply docs/source/index.md]: Done 😄

**Review by OriolAbril — APPROVED** [review]

**Comment by OriolAbril** [comment]
Thanks @symeneses! :heart:
